### ISSUE TITLE
SVM write/load problems with kernels other than linear

### ISSUE THREAD

**Issue by opencv-pushbot** [Itseez/opencv#5054]
Transferred from http://code.opencv.org/issues/4455
```
|| Jens Garstka on 2015-07-01 08:32
|| Priority: Normal
|| Affected: branch 'master' (3.0-dev)
|| Category: ml
|| Tracker: Bug
|| Difficulty: Easy
|| PR: 
|| Platform: x64 / Linux
```

SVM write/load problems with kernels other than linear
-----------
```
Within the implementation of the SVMs write and read methods (modules/ml/src/svm.cpp, lines 2022 ff and 2139 ff), the author makes the assumption, that support vector indices are not required in 2-class problems. That is probably correct if a linear kernel is used (i.e., when you have only one support vector). But with other kernels (in my case with an RBF kernel), the support vector indices are definitly required.

I've added a git diff where this questionable conditions are commented out. Without these lines everything works fine. 

IMHO the lines should be removed.
```

History
-------
##### Hyunjun Kim on 2015-07-03 07:35
```
Hi, I was able to reproduce the error. I confirmed that cv::Algorithm::load() works well only with SVM::LINEAR kernel.

<pre>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include "opencv2/imgcodecs.hpp"
#include <opencv2/highgui.hpp>
#include <opencv2/ml.hpp>

using namespace cv;
using namespace cv::ml;

int main(int, char**)
{
    // Data for visual representation
    int width = 512, height = 512;
    Mat image = Mat::zeros(height, width, CV_8UC3);

    // Set up training data
    int labels[4] = {1, -1, -1, -1};
    Mat labelsMat(4, 1, CV_32SC1, labels);

    float trainingData[4][2] = { {501, 10}, {255, 10}, {501, 255}, {10, 501} };
    Mat trainingDataMat(4, 2, CV_32FC1, trainingData);

    // Set up SVM's parameters
    Ptr<ml::SVM> svm = ml::SVM::create();
    svm->setType(ml::SVM::C_SVC);
    svm->setKernel(SVM::INTER); // Algorithm::load() works well with SVM::LINEAR
    svm->setTermCriteria(TermCriteria(TermCriteria::MAX_ITER, 100, 1e-6));

    // Train the SVM
    svm->train(ml::TrainData::create(trainingDataMat, ml::ROW_SAMPLE, labelsMat));

    // Save and load SVM
    svm->save("ex_svm.xml");
    svm= cv::Algorithm::load<ml::SVM>("ex_svm.xml"); // something is wrong


    Vec3b green(0,255,0), blue (255,0,0);
    // Show the decision regions given by the SVM
    for (int i = 0; i < image.rows; ++i)
        for (int j = 0; j < image.cols; ++j)
        {
            Mat sampleMat = (Mat_<float>(1,2) << j,i);
            float response = svm->predict(sampleMat);

            if (response == 1)
                image.at<Vec3b>(i,j)  = green;
            else if (response == -1)
                image.at<Vec3b>(i,j)  = blue;
        }

    // Show the training data
    int thickness = -1;
    int lineType = 8;
    circle( image, Point(501,  10), 5, Scalar(  0,   0,   0), thickness, lineType );
    circle( image, Point(255,  10), 5, Scalar(255, 255, 255), thickness, lineType );
    circle( image, Point(501, 255), 5, Scalar(255, 255, 255), thickness, lineType );
    circle( image, Point( 10, 501), 5, Scalar(255, 255, 255), thickness, lineType );

    // Show support vectors
    thickness = 2;
    lineType  = 8;

    Mat sv = svm->getSupportVectors();

    for (int i = 0; i < sv.rows; ++i)
    {
        const float* v = sv.ptr<float>(i);
        circle( image,  Point( (int) v[0], (int) v[1]),   6,  Scalar(128, 128, 128), thickness, lineType);
    }

    imwrite("result.png", image);        // save the image

    imshow("SVM Simple Example", image); // show it to the user
    waitKey(0);

}
</pre>
```
##### Maksim Shabunin on 2015-07-13 08:56
```
-   Assignee set to Maksim Shabunin
```

**Comment by rmekdma** [comment]
I copy the diff file here just in case.

```
diff --git a/modules/ml/src/svm.cpp b/modules/ml/src/svm.cpp
index 9318085..b9f1c51 100644
--- a/modules/ml/src/svm.cpp
+++ b/modules/ml/src/svm.cpp
@@ -2066,14 +2066,14 @@ public:
                << "alpha" << "[:";
             fs.writeRaw("d", (const uchar*)&df_alpha[df.ofs], sv_count*sizeof(df_alpha[0]));
             fs << "]";
-            if( class_count > 2 )
-            {
+//            if( class_count > 2 )
+//            {
                 fs << "index" << "[:";
                 fs.writeRaw("i", (const uchar*)&df_index[df.ofs], sv_count*sizeof(df_index[0]));
                 fs << "]";
-            }
-            else
-                CV_Assert( sv_count == sv_total );
+//            }
+//            else
+//                CV_Assert( sv_count == sv_total );
             fs << "}";
         }
         fs << "]";
@@ -2191,12 +2191,12 @@ public:
             df_index.resize(ofs + sv_count);
             df_alpha.resize(ofs + sv_count);
             dfi["alpha"].readRaw("d", (uchar*)&df_alpha[ofs], sv_count*sizeof(df_alpha[0]));
-            if( class_count > 2 )
+//            if( class_count > 2 )
                 dfi["index"].readRaw("i", (uchar*)&df_index[ofs], sv_count*sizeof(df_index[0]));
             decision_func.push_back(df);
         }
-        if( class_count <= 2 )
-            setRangeVector(df_index, sv_total);
+//        if( class_count <= 2 )
+//            setRangeVector(df_index, sv_total);
         if( (int)fn["optimize_linear"] != 0 )
             optimize_linear_svm();
     }
```

**Comment by Bardo91** [comment]
I'm not sure but, in the SVM class parameters are duplicated. There are inside svm and also inside the SVMKernelImpl class. After autotraining, only the Kernel object has the right parameters so, initially, SVM works because it uses kernel for prediction. But if I save the SVM, the wrong parameters are saved and then, if I tried to reload it for new prediction, it throws bad ones.

		cv::Ptr<cv::ml::SVM> mSvm;
		mSvm = cv::ml::SVM::create();
		mSvm->setType(ml::SVM::Types::C_SVC);
		mSvm->setKernel(ml::SVM::KernelTypes::RBF);
		mSvm->trainAuto(trainData, 10)
		mSvm->save("mode.xml");
		cv::Ptr<cv::ml::SVM> mSvm2;
		mSvm2 = cv::ml::SVM::load<cv::ml::SVM>("mode.xml");

In this example, mSvm produces good predictions but mSvm2 not.

It's possible to solve the problem by substituting `params = best_params` at the end of trainAuto(...) with `params = ((SVMKernelImpl *)&(*kernel))->params;`

**Comment by pepecarus** [comment]
I could reproduce this issue using Java OpenCV. Changed the lines of code he suggested and everything works as expected

**Comment by rmekdma** [comment]
Still not fixed in OpenCV 3.1

**Comment by Bardo91** [comment]
Is it fixed? I'm trying to reproduce it with current master and I can't.

**Comment by Bardo91** [comment]
Is it fixed? I'm trying to reproduce it with current master but I can't.  https://github.com/Bardo91/opencvSvmLoadError

**Comment by eduardohenriquearnold** [comment]
I confirm it is still **not fixed** in OpenCV 3.1 (46662ccd99acfd0fac385eb9d46e1a084f5f6ec3). I used the sample code the OP provided and it is clear that after loading the XML the SVM provides wrong results, as seen below.

With load:
![load](https://cloud.githubusercontent.com/assets/4097245/14086425/69c05e86-f4fb-11e5-9095-f003dda5aabd.png)

Without load:
![noload](https://cloud.githubusercontent.com/assets/4097245/14086424/69bbb728-f4fb-11e5-8cb7-c611e1adf9f5.png)

**Comment by eduardohenriquearnold** [comment]
As it turns out the problem was the same as described by the OP: the indexes are required for n>=2 classes when the Kernel is non-linear. Although the commented lines were removed, still, for n=2 the indexes were not being saved, which led to incorrect ordering of the decision functions. The solution was to simply enforce that even for 2 classes classifiers the indexes are saved/loaded.

I have made a patch solving this issue and will submit a push request later on.
[svm-index-order.patch.txt](https://github.com/Itseez/opencv/files/194172/svm-index-order.patch.txt)

**Comment by Goldesel23** [comment]
Is there another way to save the svm model and counter this bug? I have to load a pretrained classifier everytime the application starts... I didn't want to train the SVM everytime a new instance is started...

**Comment by mshabunin** [comment]
@Goldesel23 , did you try the fix from PR #6345 ?

**Comment by Goldesel23** [comment]
Thanks for the answer i'll take a look into it!

**Comment by animetrics** [comment]
This still seems broken for me and I'm using 3.2.0.   I did check that the patch has been applied.  save/load gives same results for linear but not for RBF.

**Comment by eduardohenriquearnold** [comment]
@animetrics I have verified that this issue has been fixed in version [3.2.0](https://github.com/opencv/opencv/archive/3.2.0.tar.gz) using the code from https://github.com/opencv/opencv/issues/5054#issuecomment-137677682, and changing the kernel to RBF. I then generated the classification zone images without loading the model (straight from training) and on a second stage I tried loading the saved model. As you can see from the results below the results are identical.

If you believe that there are still errors on this implementation please provide a code of reference showing so.

![result_noload](https://cloud.githubusercontent.com/assets/4097245/23812915/3641690c-05bb-11e7-859a-27564a218daa.png)

![result](https://cloud.githubusercontent.com/assets/4097245/23812914/36056b50-05bb-11e7-9a28-bbebd8f53457.png)

**Comment by animetrics** [comment]
@eduardohenriquearnold thanks.  I must have had some vestige of linkage to my 3.1 version.  All seems to work for me now.
